core/app/models/spree/price.rb

```diff
@@ -1,18 +1,35 @@
 module Spree
   class Price < Spree::Base
+    # We use the soft-deletion a lot, to a point where it appears we're never deleting prices.
+    # All price relations on `Spree::Variant` are scope `with_deleted`. The `valid_from` field
+    # tells us which price is currently valid.
     acts_as_paranoid
 
     MAXIMUM_AMOUNT = BigDecimal('99_999_999.99')
 
     belongs_to :variant, -> { with_deleted }, class_name: 'Spree::Variant', touch: true
 
-    validate :check_price
+    after_initialize :valid_from_today!, if: -> { valid_from.blank? }
+    after_initialize :set_default_currency, if: -> { currency.blank? }
+
     validates :amount, allow_nil: true, numericality: {
       greater_than_or_equal_to: 0,
       less_than_or_equal_to: MAXIMUM_AMOUNT
     }
 
-    after_save :set_default_price
+    scope :valid_before, -> (date) { where("#{self.table_name}.valid_from <= ?", date).order(valid_from: :desc) }
+    scope :valid_before_now, -> { valid_before(Time.current) }
+    scope :with_default_currency, -> { where(currency: Spree::Config.currency) }
+
+    # The scope to be passed into a `default_price` association for an object that needs a default price
+    # @return [ActiveRecord::Relation] A scope where the first object is the current default price
+    scope :default_price_scope, -> { with_deleted.with_default_currency.valid_before_now }
+
+    # Returns a fixed-length cache key for all prices in a passed in relation
+    #
+    def self.cache_key
+      Digest::MD5.hexdigest(pluck(:id, :updated_at).sort.flatten.join("/"))
```

Comment:
This has possible implications I can't even imagine. But maybe I'm just anxious, since `cache_key` is called very often. @jhawthorn thoughts?